src/js_native_api_v8.cc

```diff
@@ -692,6 +734,181 @@ napi_status NAPI_CDECL napi_get_last_error_info(
+
+  auto ret = node::LoadEnvironment(
+      env,
+      [main_script, env](const node::StartExecutionCallbackInfo& info)
+          -> v8::MaybeLocal<v8::Value> {
+        node::Realm* realm = env->principal_realm();
+        auto ret = realm->ExecuteBootstrapper(
+            "internal/bootstrap/switches/is_embedded_env");
+        if (ret.IsEmpty()) return ret;
+
+        std::string name =
+            "embedder_main_napi_" + std::to_string(env->thread_id());
+        env->builtin_loader()->Add(name.c_str(), main_script);
+        return realm->ExecuteBootstrapper(name.c_str());
+      });
+  if (ret.IsEmpty()) return napi_pending_exception;
+
+  emb_env->seal();
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_destroy_environment(napi_env env, int* exit_code) {
+  CHECK_ARG(env, env);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  int r = node::SpinEventLoop(node_env->node_env()).FromMaybe(1);
+  if (exit_code != nullptr) *exit_code = r;
+  node::Stop(node_env->node_env());
+
+  auto emb_env = reinterpret_cast<v8impl::EmbeddedEnvironment*>(
+      node_env->node_env()->get_embedded());
+  node_env->node_env()->set_embedded(nullptr);
+  // This deletes the uniq_ptr to node::CommonEnvironmentSetup
+  // and the v8::locker
+  delete emb_env;
+
+  return napi_ok;
+}
+
+napi_status NAPI_CDECL napi_run_environment(napi_env env) {
+  CHECK_ARG(env, env);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  if (node::SpinEventLoopWithoutCleanup(node_env->node_env()).IsNothing())
+    return napi_closing;
+
+  return napi_ok;
+}
+
+static void napi_promise_error_handler(
+    const v8::FunctionCallbackInfo<v8::Value>& info) {
+  return;
+}
+
+napi_status napi_await_promise(napi_env env,
+                               napi_value promise,
+                               napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(env, result);
+
+  v8::EscapableHandleScope scope(env->isolate);
+  node_napi_env node_env = reinterpret_cast<node_napi_env>(env);
+
+  v8::Local<v8::Value> promise_value = v8impl::V8LocalValueFromJsValue(promise);
+  if (promise_value.IsEmpty() || !promise_value->IsPromise())
+    return napi_invalid_arg;
+  v8::Local<v8::Promise> promise_object = promise_value.As<v8::Promise>();
+
+  v8::Local<v8::Value> rejected = v8::Boolean::New(env->isolate, false);
+  v8::Local<v8::Function> err_handler =
+      v8::Function::New(env->context(), napi_promise_error_handler, rejected)
+          .ToLocalChecked();
+
+  if (promise_object->Catch(env->context(), err_handler).IsEmpty())
+    return napi_pending_exception;
+
+  if (node::SpinEventLoopWithoutCleanup(
+          node_env->node_env(),
+          [&promise_object]() {
+            return promise_object->State() ==
+                   v8::Promise::PromiseState::kPending;
+          })
+          .IsNothing())
+    return napi_closing;
+
+  *result =
+      v8impl::JsValueFromV8LocalValue(scope.Escape(promise_object->Result()));
```

Comment:
`SpinEventLoopWithoutCleanup` can return for either of two reasons:
  1. The condition lambda returns false (because the promise is resolved)
  2. There are no pending callbacks.

In the second case, the call to `promise_object->Result()` here may terminate the process with `SIGABRT` and the following error message:
```
FATAL ERROR: v8_Promise_Result Promise is still pending
```

I encountered that error from passing in a promise with an empty resolve callback -- my intention was to just have it run indefinitely for now, and I'd consider graceful cleanup later.

Should this code check the promise state again, and return a different error code (instead of aborting the process) if the promise is not resolved?

Reply:
Node.js in general has an absolutely horrible behavior when it comes to `Promise`s that cannot be resolved because they have no handlers - it always simply quits - without an error code. This has already been discussed and it seems that it will stay like this. Anyway, the other possible valid behavior is to simply hang forever. So, in this case, besides suppressing the error message and the error code, there isn't much that can be done.